File: Dataset/Leetcode/valid/13/194.java
class Solution {
    public int XXX(String s) {
        int sum = 0;
        HashMap<String, Integer> hashMap = new HashMap<>();
        hashMap.put("I", 1);
        hashMap.put("IV", 4);
        hashMap.put("V", 5);
        hashMap.put("IX", 9);
        hashMap.put("X", 10);
        hashMap.put("XL", 40);
        hashMap.put("L", 50);
        hashMap.put("XC", 90);
        hashMap.put("C", 100);
        hashMap.put("CD", 400);
        hashMap.put("D", 500);
        hashMap.put("CM", 900);
        hashMap.put("M", 1000);
        int i;
        for (i = 0; i < s.length() - 1; i++) {
            String s1 = s.substring(i, i + 1);
            if ("I".equals(s1) || "X".equals(s1) || "C".equals(s1)) {
                String s2 = s.substring(i, i + 2);
                 if (hashMap.containsKey(s2)) {
                     sum = sum + hashMap.get(s2);
                     i++;
                     continue;
                 }
            }
            sum = sum + hashMap.get(s1);
        }
        if (i < s.length()) {
            sum = sum + hashMap.get(s.substring(s.length() - 1));

        }
        return sum;
    }
}

File: Dataset/Leetcode/train/62/194.cpp
 class Solution {
public:
    int XXX(int m, int n) {
        //最朴素的做法
        /*
        vector<vector<int>> dp(m, vector<int>(n, 0));
        for (int i = 0; i < m; i++) {
            dp[i][0] = 1;
        }
        for (int j = 1; j < n; j++) {
            dp[0][j] = 1;
        }

        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                dp[i][j] = dp[i-1][j] + dp[i][j-1];
            }
        }
        return dp[m-1][n-1];
        */
        //分析以后可以看出只需要一维数组就可以解决
        //分析就是当前格子的方法数 等于上一行格子的加上左边格子的，我们用一维数组来存储上一行的值，又因为左边都是从1开始
        vector<int> dp(n, 0);
        dp[0] = 1;
        for (int i = 0; i < m; i++) {
            for (int j = 1; j < n; j++) {
                dp[j] = dp[j] + dp[j-1];
            }
        }
        return dp[n-1];
    }
};

